python.md

```diff
@@ -83,6 +84,18 @@ with the convention unless you have a very good reason not to.
 This convention also applies to variable names.
 
 
+### Model method naming conventions
+
+When sorting is involved, here are some guidelines to keep things consistent:
+
+- Prefer "active" to "current", ie `get_active_direct_debit_instruction`
+
+- Prefer "latest" to "last" as "latest" implies chronological order where the
+  ordering has to be looked up when using "last"; ie
+  `get_latest_payment_schedule`. Similarly, prefer `earliest` to `first` in
+  method names.
```

Comment:
Any more conventions we should document?

Reply:
Maybe some guidance about to prefix a method with `get_`? Should every model method (as opposed to inexpensive properties) use that prefix?

Reply:
Also, I wonder if we should revise [the CharField choices guidance](https://github.com/octoenergy/styleguide/blob/faf9c931c42a7e1d71655228984fe301d2653632/python.md#charfield-choices). It's safer to prefix every choice constant with the name of the field to avoid collisions. Take, for example, the `source` and `preferred_communication_method` fields on `ComplaintContact`. These follow the conventions, but actually share some choice names, so the constants are defined twice. ([See here](https://github.com/octoenergy/consumer-site/blob/master/src/octoenergy/support/models.py#L573)).

Instead, maybe the example should use `CHANNEL_TELESALES` and `CHANNEL_FIELD_SALES`.

Reply:
Another thing: should we discourage the use of `auto_now` and `auto_now_add`? These can be a pain to work with in testing, as it will silently discard the datetime that is passed to the field. If we want to do this at the model level at all then I think`default=timezone.now` is a better approach, as it allows overriding. 
 See [the docs](https://docs.djangoproject.com/en/2.0/ref/models/fields/#django.db.models.DateField.auto_now_add) and also [this discussion](https://code.djangoproject.com/ticket/22995).

Reply:
> Also, I wonder if we should revise the CharField choices guidance. It's safer to prefix every choice constant with the name of the field to avoid collisions. Take, for example, the source and preferred_communication_method fields on ComplaintContact. These follow the conventions, but actually share some choice names, so the constants are defined twice. (See here).

Yes, we should definitely do that - I'll update this PR.

> Another thing: should we discourage the use of auto_now and auto_now_add? These can be a pain to work with in testing, as it will silently discard the datetime that is passed to the field. If we want to do this at the model level at all then I thinkdefault=timezone.now is a better approach, as it allows overriding.

Seems sensible. You can work around the testing issue by overriding `_create` on the factory  boy class but that's not a strong enough reason to stick with current convention.

Can we easily convert existing classes to use the `default=` approach?

Reply:
> Can we easily convert existing classes to use the default= approach?

Yes, I believe you just change it and then make/run migrations. There is no effect on the database.

Reply:
I personally think of fields that use `auto_now` and `auto_now_add` to be audit fields, rather than actual "data" fields. They signal (in Django's world only) when an action of creating or updating a model occurred, which is not always the same as the related data. I think a `created_at` and `updated_at` fields for auditing using these kwargs are good and should be set and forget, and then use explicit fields for related to the data of the instance. e.g.

```
class Account(models.Model):
    joined_at = models.DateTimeField()
    status_updated_at = models.DateTimeField(blank=True, null=True)

    # Audit fields
    created_at = models.DateTimeField(auto_now_add=True)
    updated_at = models.DateTimeField(auto_now=True)    
```

Reply:
@jamesbeith Are are you saying application logic should never depend on `created_at` fields (and so you would never want to manipulate them in tests)?

Reply:
I would tend to agree with @jamesbeith's thinking on these fields. I regard them almost as 'a feature which should be provided by the database, but isn't so we have to add it ourselves'. I find them very useful as an inviolable record of what *actually* happened in the lifecycle of a row. If they cease to mean that, because they can be overridden, I would want to know.

Also, for `auto_now_add` in particular, there's a fair bit more work involved, as you need to override `save` so that it only sets the value to `timezone.now` when *creating* the model instance, and leaves it thereafter.

Reply:
Yeah, that’s right; "application logic" is a good way to phrase it.

These two "audit fields" should only be thought of as when Django created/updated the instance. If the application logic needs to know when something happened, e.g. joined, enrolled, etc., then this should be its own distinct field on the model (without `auto_now(_add)`. The value for such fields should be created/received at the interface layer and passed all the way down to the data layer (via domain), rather than using `default=timezone.now` (something you’ve suggested recently which I like).

Reply:
Apologies, my mistake. It's `auto_now` which needs the override, not `auto_now_add`. That is to say, `default` will set the value of `updated_at` to `timezone.now` when the object is *created*, but not subsequently. So `updated_at` will become stale the second time the object is saved, unless it is manually provided by the caller. `auto_now` will set `updated_at` to the current timezone *every time* the object is saved. This is where we would need to override `save` to get the same behaviour.

Reply:
Ah I see, yes that makes sense. I like the idea of audit fields have a particular pattern (could we even encapsulate them into a mixin?

Incidentally [Django Audit Log](https://github.com/vvangelovski/django-audit-log) was what we used at our last place, might be worth considering at some point.